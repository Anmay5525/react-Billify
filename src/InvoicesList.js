import Table from "./Table.js";
import Loader from "./Loader.js";

const { useEffect, useState } = React;

const InvoicesList = () => {
  const [invoices, setInvoices] = useState(null);

  const getInvoices = () => {
    fetch("https://rzp-training.herokuapp.com/team1/invoices")
      .then((res) => res.json())
      .then((r) => setInvoices(r))
      .catch((er) => console.log(er));
  };
  useEffect(getInvoices, []);
  const fields = ["date", "customer", "status", "amount", "amount_due"];
  if (invoices) {
    const data = invoices.items.map((item) => {
      return {
        date: new Date(item.date * 1000).toDateString(),
        customer: item.customer_details.name,
        status: item.status.toUpperCase(),
        amount: item.amount / 100,
        amount_due: item.amount_due / 100,
      };
    });
    return (
      <div className="content">
        <div className="invoices-title-container">
          <div className="invoices-title">Invoices</div>
          <button className="invoices-new-btn">+ New Invoice</button>
        </div>
        <Table fields={fields} data={data} />
      </div>
    );
  } else {
    return <Loader />;
  }
};
export default InvoicesList;
