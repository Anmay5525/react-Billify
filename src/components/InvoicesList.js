import React, { useEffect, useState } from "react";
import Table from "./Table";
import Loader from "./Loader";

// const { useEffect, useState } = React;

const InvoicesList = () => {
  const [invoices, setInvoices] = useState(null);

  const controller = new AbortController();
  const { signal } = controller;

  const getInvoices = () => {
    fetch("https://rzp-training.herokuapp.com/team1/invoices", { signal })
      .then((res) => res.json())
      .then((r) => setInvoices(r))
      // eslint-disable-next-line no-console
      .catch((er) => console.log(er));
  };
  useEffect(() => {
    getInvoices();
    return () => {
      controller.abort();
    };
  }, []);
  const fields = ["date", "customer", "status", "amount", "amount_due"];
  if (invoices) {
    const data = invoices.items.map((item) => {
      return {
        date: new Date(item.date * 1000).toDateString(),
        customer: item.customer_details.name,
        status: item.status.toUpperCase(),
        amount: item.amount / 100,
        amount_due: item.amount_due / 100,
      };
    });
    return (
      <div className="content">
        <div className="invoices-title-container">
          <div className="invoices-title">Invoices</div>
          <button className="invoices-new-btn">+ New Invoice</button>
        </div>
        <Table fields={fields} data={data} />
      </div>
    );
  }
  return <Loader />;
};
export default InvoicesList;
