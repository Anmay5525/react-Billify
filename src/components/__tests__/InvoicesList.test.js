import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import InvoicesList from '../InvoicesList';
import { BrowserRouter as Router } from "react-router-dom";

const invoicesListData = {
  "entity": "collection",
  "count": 1,
  "items": [
    {
      "id": "inv_Fg1JWqxU28xdI6",
      "entity": "invoice",
      "receipt": null,
      "invoice_number": null,
      "customer_id": "cust_FfdAl5kGR1Rk4v",
      "customer_details": {
        "id": "cust_FfdAl5kGR1Rk4v",
        "name": "Parmanad",
        "email": "parmanand@example.com",
        "contact": "9920391003",
        "gstin": null,
        "billing_address": null,
        "shipping_address": null,
        "customer_name": "Parmanad",
        "customer_email": "parmanand@example.com",
        "customer_contact": "9920391003"
      },
      "order_id": "order_Fg1JWtRICHHnz8",
      "line_items": [
        {
          "id": "li_Fg1JWrtQHCafM9",
          "item_id": "item_Ffd9vOPr1yKiUC",
          "ref_id": null,
          "ref_type": null,
          "name": "disc",
          "description": "disc",
          "amount": 30000,
          "unit_amount": 30000,
          "gross_amount": 30000,
          "tax_amount": 0,
          "taxable_amount": 30000,
          "net_amount": 30000,
          "currency": "INR",
          "type": "invoice",
          "tax_inclusive": false,
          "hsn_code": null,
          "sac_code": null,
          "tax_rate": null,
          "unit": null,
          "quantity": 1,
          "taxes": []
        }
      ],
      "payment_id": null,
      "status": "issued",
      "expire_by": 1601251200,
      "issued_at": 1600765613,
      "paid_at": null,
      "cancelled_at": null,
      "expired_at": null,
      "sms_status": "sent",
      "email_status": "sent",
      "date": 1600992000,
      "terms": null,
      "partial_payment": false,
      "gross_amount": 30000,
      "tax_amount": 0,
      "taxable_amount": 30000,
      "amount": 30000,
      "amount_paid": 0,
      "amount_due": 30000,
      "currency": "INR",
      "currency_symbol": "â‚¹",
      "description": null,
      "notes": [],
      "comment": "akd",
      "short_url": "https://rzp.io/i/ioJMd5r",
      "view_less": true,
      "billing_start": null,
      "billing_end": null,
      "type": "invoice",
      "group_taxes_discounts": false,
      "created_at": 1600765613,
      "idempotency_key": null
    }
  ]
};

// it('makes fetch calls', async (done) => {
//   fetch.mockResponse(JSON.stringify(invoiceListData));
//   // await act(async () => {container = mount(<Router><InvoicesList/></Router>)});
//   // expect(container.debug()).toMatchSnapshot();
//   let container;
//   await act(async() => container = mount(<Router><InvoicesList/></Router>));
//   setTimeout(() => {
//     // expectations here
//     expect(container.html()).toMatchSnapshot();
//     done();
//   }, 500);
// });

describe('Testing InvoicesList component if it', () => {

  beforeEach(() => {
    fetch.resetMocks();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<InvoicesList></InvoicesList>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders loader', () => {
    const tree = renderer.create(<InvoicesList></InvoicesList>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('makes fetch calls', async () => {
    fetch.mockResponseOnce(JSON.stringify(invoicesListData));
    const container = mount(<Router><InvoicesList /></Router>);
    await act(async () => { });
    container.update();
    expect(container.html()).toMatchSnapshot();
    expect(fetch).toHaveBeenCalledTimes(1);
  });

  it('handles fetch errors', async () => {
    fetch.mockRejectOnce(['Error', {status: 200}]);
    const container = mount(<Router><InvoicesList /></Router>);
    await act(async () => { });
    container.update();
    expect(container.html()).toMatchSnapshot();
  });

  it('displays correct button type', async () => {
    fetch.mockResponseOnce(JSON.stringify(invoicesListData));
    const container = mount(<Router><InvoicesList /></Router>);
    await act(async () => { });
    container.update();
    expect(container.find('button').at(0).type()).toEqual('button');
  });

  it('displays correct table columns', async () => {
    fetch.mockResponseOnce(JSON.stringify(invoicesListData));
    const container = mount(<Router><InvoicesList /></Router>);
    await act(async () => { });
    container.update();
    const tr0 = container.find('tr').at(0);
    expect(tr0.childAt(0).contains('DATE')).toEqual(true);
    expect(tr0.childAt(1).contains('CUSTOMER')).toEqual(true);
    expect(tr0.childAt(2).contains('STATUS')).toEqual(true);
    expect(tr0.childAt(3).contains('AMOUNT')).toEqual(true);
    expect(tr0.childAt(4).contains('AMOUNT DUE')).toEqual(true);
  });

  it('displays correct table data', async () => {
    fetch.mockResponseOnce(JSON.stringify(invoicesListData));
    const container = mount(<Router><InvoicesList /></Router>);
    await act(async () => { });
    container.update();
    const tr1 = container.find('tr').at(1);
    expect(tr1.childAt(0).contains('Fri Sep 25 2020')).toEqual(true);
    expect(tr1.childAt(1).contains('Parmanad')).toEqual(true);
    expect(tr1.childAt(2).contains('ISSUED')).toEqual(true);
    expect(tr1.childAt(3).contains(300)).toEqual(true);
    expect(tr1.childAt(4).contains(300)).toEqual(true);
  });

});
