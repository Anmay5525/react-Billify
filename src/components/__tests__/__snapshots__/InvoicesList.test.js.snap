// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing InvoicesList component if it handles fetch errors 1`] = `"<div class=\\"loader-container\\"><svg viewBox=\\"0 0 100 100\\" xmlns=\\"http://www.w3.org/2000/svg\\"><circle cx=\\"50\\" cy=\\"50\\" r=\\"45\\"></circle></svg></div>"`;

exports[`Testing InvoicesList component if it makes fetch calls 1`] = `"<div class=\\"content\\"><div class=\\"invoices-title-container\\"><div class=\\"invoices-title\\">Invoices</div><a href=\\"/Invoices/New\\"><button type=\\"button\\" class=\\"invoices-new-btn\\">+ New Invoice</button></a></div><div><table class=\\"table\\"><thead><tr><th>DATE</th><th>CUSTOMER</th><th>STATUS</th><th>AMOUNT</th><th>AMOUNT DUE</th></tr></thead><tbody><tr class=\\"tr\\"><td>Fri Sep 25 2020</td><td>Parmanad</td><td>ISSUED</td><td>300</td><td>300</td></tr></tbody></table></div></div>"`;

exports[`Testing InvoicesList component if it renders loader 1`] = `
<div
  className="loader-container"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="50"
      cy="50"
      r="45"
    />
  </svg>
</div>
`;
